### PR TITLE
Added base subject class

### PR DESCRIPTION
**Pull request by AlexandrGraschenkov**
I restore back mutex because it less expensive for usage than thread. Deadlock problem developer can solve with add observer dispatch queue.

Subject sometimes can be usefull. E.g. when we call something from net, and waiting result. For this moment you don't have value.

### CODE REVIEW

**Comment by roberthein** [comment]
I guess this one is obselete now @AlexandrGraschenkov ?

**Comment by AlexandrGraschenkov** [comment]
@roberthein you talking about swift 4? If so, I can update to swift 5 version.

**Comment by roberthein** [comment]
What I mean @AlexandrGraschenkov is that Mutex was already restored in another PR, sorry for that..

**Comment by 4brunu** [comment]
This PR worries me a bit, because the `value` becomes optional, and in some subclasses there are some force unwrap to deal with it...

File: Observable/Classes/Base.swift [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    
+    public typealias Observer = (T, T?) -> Void
+    
+    private var observers: [Int: (Observer, DispatchQueue?)] = [:]
+    private var uniqueID = (0...).makeIterator()
+    
+    let lock: Lock = Mutex()
+    var _value: T? {
+        didSet {
+            let val = _value!
```
Comment (4brunu) [inline Observable/Classes/Base.swift]: If the _value becomes nil, which is a valid option, this will crash

Comment (4brunu) [inline Observable/Classes/Base.swift:13]: If the _value becomes nil, which is a valid option, this will crash

Comment (AlexandrGraschenkov) [reply Observable/Classes/Base.swift:13]: Set `nil` value possible only inside. It is internal property. Outside it's not available.

**Comment by AlexandrGraschenkov** [comment]
@roberthein there also some changes inside, 2 of them already accepted:
- [x] reset mutex back #22 
- [x] #26 + some other cases, when we need thread safe
- [ ] added base subject class; It's for case when we don't received value yet

There also lot of file changed, not contains code. So PR little dirty, I can remove not depended changes from PR.

File: Observable/Classes/Base.swift (hunk 1 above)
Comment (4brunu) [reply Observable/Classes/Base.swift:13]: I have tested it, and you are right.

**Comment by 4brunu** [comment]
Can you also merge this PR https://github.com/roberthein/Observable/pull/25?

**Comment by AlexandrGraschenkov** [comment]
@4brunu ok, today will try include PR #25

**Comment by AlexandrGraschenkov** [comment]
Sorry for 2 weeks waiting. TG contest is finished, now I found time to resolve all thing in this repo.

**Pull request closed without merging**